src/ghga_connector/core/downloading/batch_processing.py

```diff
@@ -129,167 +129,108 @@ def handle_response(self, *, response: str):
+    def __init__(  # noqa: PLR0913
         self,
-        api_url: str,
-        file_ids_with_extension: dict[str, str],
+        *,
+        wanted_file_ids: list[str],
+        dcs_api_url: str,
+        output_dir: Path,
         max_wait_time: int,
-        retry_after: int = 60,
-    ) -> None:
-        self.api_url = api_url
-        self.check_api_available()
-        self.file_ids_with_extension = file_ids_with_extension
+        message_display: AbstractMessageDisplay,
+        work_package_accessor: WorkPackageAccessor,
+    ):
+        """Initialize the FileStager."""
+        io_handler = CliIoHandler()
+        existing_file_ids = set(io_handler.check_output(location=output_dir))
+        if not is_service_healthy(dcs_api_url):
+            raise exceptions.ApiNotReachableError(api_url=dcs_api_url)
+        self.api_url = dcs_api_url
+        self.message_display = message_display
+        self.work_package_accessor = work_package_accessor
         self.max_wait_time = max_wait_time
-        # amount of seconds between staging attempts
-        self.retry_after = retry_after
-
-    def check_api_available(self):
-        """Get response from endpoint, else throw corresponding exception"""
-        if not is_service_healthy(self.api_url):
-            raise exceptions.ApiNotReachableError(api_url=self.api_url)
-
-    def get_file_ids(self) -> list[str]:
-        """Get file ids that should be staged"""
-        return list(self.file_ids_with_extension.keys())
-
-    def remove_existing(self, *, file_ids: list[str]):
-        """Delete file information for already existing files."""
-        for file_id in file_ids:
-            del self.file_ids_with_extension[file_id]
-
-
-@dataclass
-class StagingState:
-    """Handle state for staged and not yet staged file ids."""
-
-    time_started: datetime = field(default_factory=datetime.utcnow, init=False)
-    staged_files: list[str] = field(default_factory=list, init=False)
-    unstaged_files: list[str] = field(default_factory=list, init=False)
-
-    def add_staged(self, *, file_id: str):
-        """Delegate adding staged file ids"""
-        self.staged_files.append(file_id)
-
-    def add_unstaged(self, *, file_id: str):
-        """Delegate adding unstaged file ids"""
-        self.unstaged_files.append(file_id)
-
-    def update_staged_files_wait(
-        self, *, max_wait_time: int, work_package_accessor: WorkPackageAccessor
-    ) -> bool:
-        """
-        Update staged file list after all previously staged files have been processed.
-        Caller has to make sure, that all file ids in self.staged_files have actually
-        been processed.
-
-        Returns a boolean instructing the caller to wait for retry_time and call this
-        method again.
-        """
-        self.staged_files = []
-        remaining_unstaged = []
-
-        for file_id in self.unstaged_files:
+        self.time_started = now = time()
+        self.staged = {}
+        self.unstaged = {
```

Comment:
Since the previous version used just a list of file IDs, I think it would be good to adjust the names of these, leave a comment, or type-hint to suggest the values of the dicts. `unstaged_retry_times` or something along those lines.

Reply:
Similarly, `file_staging_responses`, `staged_file_responses`, `staged_file_to_response`

Reply:
I had added this as a comment where the attribute was defined. Was reluctant to make the name longer because the caller actually uses only the keys, not the values. But you're probably right, better to give it a more descriptive name. Changed in b7ded70 as well.